lib/internal/modules/esm/loader.js

```diff
@@ -202,33 +211,74 @@ class ESMLoader {
       const { ModuleWrap, callbackMap } = internalBinding('module_wrap');
       const module = new ModuleWrap(url, undefined, source, 0, 0);
       callbackMap.set(module, {
-        importModuleDynamically: (specifier, { url }) => {
-          return this.import(specifier, url);
+        importModuleDynamically: (specifier, { url }, importAssertions) => {
+          return this.import(specifier, url, importAssertions);
         }
       });
 
       return module;
     };
     const job = new ModuleJob(this, url, evalInstance, false, false);
     this.moduleMap.set(url, job);
+    finalFormatCache.set(job, 'module');
     const { module } = await job.run();
 
     return {
       namespace: module.getNamespace(),
     };
   }
 
-  async getModuleJob(specifier, parentURL) {
+  async getModuleJob(specifier, parentURL, importAssertions) {
+    if (!ArrayPrototypeIncludes(supportedTypes, importAssertions.type)) {
+      throw new ERR_INVALID_IMPORT_ASSERTION('type', importAssertions.type);
+    }
+
     const { format, url } = await this.resolve(specifier, parentURL);
     let job = this.moduleMap.get(url);
     // CommonJS will set functions for lazy job evaluation.
     if (typeof job === 'function') this.moduleMap.set(url, job = job());
 
-    if (job !== undefined) return job;
+    if (job != null) {
+      const currentImportAssertionType = importAssertionTypeCache.get(job);
+      if (currentImportAssertionType === importAssertions.type) return job;
+
+      try {
+        // To avoid race conditions, wait for previous module to fulfill first.
+        await job.modulePromise;
+      } catch {
+        // If the other job failed with a different `type` assertion, we got
+        // another chance.
+        job = undefined;
+      }
+
+      if (job !== undefined) {
+        const finalFormat = finalFormatCache.get(job);
+        if (importAssertions.type == null && finalFormat === 'json') {
+          throw new ERR_MISSING_IMPORT_ASSERTION(url, finalFormat,
+                                                 'type', 'json');
+        }
+        if (
+          importAssertions.type == null ||
+              (importAssertions.type === 'json' && finalFormat === 'json')
+        ) return job;
+        throw new ERR_FAILED_IMPORT_ASSERTION(
+          url, 'type', importAssertions.type, finalFormat);
+      }
+    }
```

Comment:
I would think that this whole block could move into `import` after the module job is returned to that function. Also I’m not sure why it’s `!= null` rather than the original `!== undefined`.